### PR TITLE
Fix/kldiv

### PR DESCRIPTION
**Pull request by S1ro1**
## Summary
Fixes KLDiv being calculated wrong due to invalid tests as described in #259 . Now made KLDiv work for even bigger vocab size etc. The benchmarks remain +- the same. Therefore not adding those as I don't have access to the original GPU they were done on before.

## Testing Done
Tests stay the same, except replacing `assert_verbose_all_close` with `assert torch.allclose`

- Hardware Type: Nvidia RTX A5500
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/ops/kl_div.py [hunk 1]
```diff
@@ -65,16 +66,18 @@ def _kldiv_kernel_forward(
         # KL(y_true || y) = y_true * (log(y_true) - log(y))
         # We compute KL(y_true || y) with y in the log-space
         if not log_target:
-            loss = y_true * (tl.log(y_true) - y)
+            eps = 1e-10
```
Comment (lancerts) [inline src/liger_kernel/ops/kl_div.py:69]: lets pass eps as a function arg.

File: test/transformers/test_kl_div.py [hunk 2]
```diff
@@ -85,12 +83,12 @@ def _test_correctness_once(
 
     output.backward()
     output2.backward()
-    assert_verbose_allclose(x1.grad, x2.grad, atol=atol, rtol=rtol)
+    assert torch.allclose(x1.grad, x2.grad, atol=atol, rtol=rtol)
 
 
 @pytest.mark.parametrize(*_SHAPE_PARAMS)
-@pytest.mark.parametrize("log_target", [False, True])
```
Comment (lancerts) [inline test/transformers/test_kl_div.py:92]: these changes seems identical and has no effect?

Comment (S1ro1) [reply test/transformers/test_kl_div.py:92]: Oh ye, those were just me reordering them to match between both calls because of ocd and probably staged it by mistake.

File: src/liger_kernel/ops/kl_div.py (hunk 1 above)
Comment (S1ro1) [reply src/liger_kernel/ops/kl_div.py]: I suppose we want to propagate it as an argument to the `transformers/LigerKLDiv..` class as well, right?

**Comment by lapp0** [comment]
Looks good, however do you have any idea why aggregation is different? I have to divide by the last dimension to get torch's `kl_div` result

```
>>> torch.nn.functional.kl_div(student_log_prob, teacher_prob).item()                                                                                                                                                                                                       
0.0004941617953591049
```

```
>>> liger_kl_div(student_log_prob, teacher_prob).item()
24.289039611816406
liger_kl_div(student_log_prob, teacher_prob).item() / teacher_prob.shape[-1]
0.0004941617759565512
```

**Comment by S1ro1** [comment]
> lapp0

It's because by default torch uses reduction `mean` and throws a warning that you should `batchmean` to get the true KL Divergence, here `batchmean` is used by default.

https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html

**Comment by S1ro1** [comment]
@lancerts Should reference all the comments and good to be merged

**Comment by lapp0** [comment]
I'm seeing consistent loss, but inconsistent gradients. Do you know if I'm doing something wrong here, or is there an issue with gradient computation?

```
KL Divergence Loss: 0.7544127702713013
Input logits_p: -1.0672125816345215 (checksum)
Input logits_q: 14.455729484558105 (checksum)
Gradients of logits_p (batch item 0): tensor([-0.0129,  0.0021, -0.0077,  0.0011, -0.0121, -0.0032, -0.0014,  0.0038],
       device='cuda:0')

LIGER:
KL Divergence Loss: 0.7544127106666565
Input logits_p: -1.0672125816345215 (checksum)
Input logits_q: 14.455729484558105 (checksum)
Gradients of logits_p (batch item 0): tensor([0.0835, 0.1133, 0.1089, 0.1140, 0.0780, 0.0999, 0.1121, 0.1176],
       device='cuda:0')
```

Code:

```
from functools import partial
import torch
import torch.nn.functional as F
from liger_kernel.transformers.functional import liger_kl_div


def run_kl_div_experiment(logits_p, logits_q, kl_fn):
    logits_p = logits_p.clone().detach().requires_grad_(True)
    p = F.log_softmax(logits_p, dim=0)
    q = F.softmax(logits_q, dim=0)

    kl_loss = kl_fn(p, q)
    kl_loss.backward()

    print(f"KL Divergence Loss: {kl_loss.item()}")
    print(f"Input logits_p: {logits_p.sum()} (checksum)")
    print(f"Input logits_q: {logits_q.sum()} (checksum)")
    print(f"Gradients of logits_p (batch item 0): {logits_p.grad[0]}")

torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")


run_kl_div_experiment(logits_p, logits_q, kl_fn=partial(F.kl_div, reduction="batchmean"))

print()
print("LIGER:")
run_kl_div_experiment(logits_p, logits_q, kl_fn=liger_kl_div)
```

**Comment by S1ro1** [comment]
I Managed to make an example similar to yours work, I don't yet know why the example you provided doesn't work, you had softmax/log_softmax applied across wrong dimension but that wasn't the primary issue so fixed that. It appears that applying softmax after detaching does something and torch/liger interprets it different way, feels like I'm missing something very trivial.

```python
from functools import partial
import torch
import torch.nn.functional as F
from liger_kernel.transformers.functional import liger_kl_div


def run_kl_div_experiment(logits_p, logits_q, kl_fn):
    p = logits_p.detach().clone().requires_grad_(True)
    q = logits_q.detach().clone()

    kl_loss = kl_fn(p, q)
    kl_loss.backward()

    print(f"KL Divergence Loss: {kl_loss.item()}")
    print(f"Input logits_p: {logits_p.sum()} (checksum)")
    print(f"Input logits_q: {logits_q.sum()} (checksum)")
    print(f"Gradients of logits_p (batch item 0): {p.grad[0]}")


torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda").log_softmax(dim=-1)
logits_q = torch.randn(8, 8, device="cuda").softmax(dim=-1)

run_kl_div_experiment(
    logits_p, logits_q, kl_fn=partial(F.kl_div, reduction="batchmean")
)

print()
print("LIGER:")
run_kl_div_experiment(logits_p, logits_q, kl_fn=liger_kl_div)
```

**Comment by lapp0** [comment]
Sorry for getting dimension wrong, was just throwing together a quick reproducer. 

It's not due to `.detach()` per the below script:

```
from functools import partial
import torch
import torch.nn.functional as F
from liger_kernel.transformers.functional import liger_kl_div


def run_kl_div_experiment(logits_p, logits_q, kl_fn):
    print("logits_q[0]", logits_q[0])
    kl_loss = kl_fn(
        logits_p.log_softmax(dim=-1), 
        logits_q.softmax(dim=-1)
)
    kl_loss.backward()

    print(f"Gradients of logits_p {logits_p.grad}")

print("TORCH:")
torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")
run_kl_div_experiment(logits_p, logits_q, kl_fn=partial(F.kl_div, reduction="batchmean"))

print("LIGER:")
torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")
run_kl_div_experiment(logits_p, logits_q, kl_fn=liger_kl_div)

```

Output:

```
logits_q[0] tensor([ 0.1808, -0.5523,  0.9238, -0.7350,  1.3800,  0.8676,  0.1297, -0.9406],
       device='cuda:0')
Gradients of logits_p tensor([[-0.0019,  0.0107, -0.0231,  0.0243, -0.0170, -0.0095,  0.0022,  0.0142],
        [-0.0252, -0.0250, -0.0032,  0.0520, -0.0117, -0.0132,  0.0166,  0.0098],
        [ 0.0030, -0.0128,  0.0373, -0.0114, -0.0054, -0.0022, -0.0045, -0.0040],
        [ 0.0021,  0.0038,  0.0285,  0.0103, -0.0090, -0.0101, -0.0209, -0.0048],
        [ 0.0044,  0.0073, -0.0626,  0.0026,  0.0039,  0.0092, -0.0014,  0.0366],
        [ 0.0046,  0.0074,  0.0375,  0.0125, -0.0658,  0.0164, -0.0054, -0.0072],
        [ 0.0091,  0.0159, -0.0189,  0.0050,  0.0292, -0.0593, -0.0002,  0.0191],
        [ 0.0035,  0.0010, -0.0289,  0.0257,  0.0136, -0.0072,  0.0040, -0.0118]],
       device='cuda:0')
LIGER:
logits_q[0] tensor([ 0.1808, -0.5523,  0.9238, -0.7350,  1.3800,  0.8676,  0.1297, -0.9406],
       device='cuda:0')
Gradients of logits_p tensor([[ 0.1019,  0.1138,  0.0776,  0.1156,  0.0521,  0.0800,  0.1030,  0.1173],
        [ 0.0731,  0.0695,  0.1163,  0.1084,  0.0861,  0.0894,  0.1129,  0.1042],
        [ 0.1140,  0.0416,  0.1186,  0.0520,  0.1030,  0.1161,  0.1069,  0.1120],
        [ 0.1133,  0.1179,  0.0839,  0.1159,  0.0942,  0.0989,  0.0342,  0.1045],
        [ 0.1199,  0.1213,  0.0049,  0.1189,  0.1206,  0.1192,  0.0805,  0.0841],
        [ 0.1163,  0.1120,  0.1093,  0.1139, -0.0036,  0.1085,  0.1103,  0.1027],
        [ 0.1229,  0.1210,  0.0559,  0.1126,  0.1235,  0.0121,  0.0980,  0.1235],
        [ 0.1156,  0.1067,  0.0556,  0.1062,  0.1145,  0.0804,  0.0885,  0.0929]],
       device='cuda:0')
```

**Comment by S1ro1** [comment]
The issue is the kernel doesn't work correctly when there is computation after the detach, i.e. log_softmax in your case, just going through the code I see the same issue in `cross_entropy_loss` where modifying the input tensor in place interferes with torch autograd engine. Would need confirmation from someone whether my thoughts are correct

**Comment by S1ro1** [comment]
@ByronHsu could you check this one quickly? If what I'm saying is true, cross_entropy, possibly fused_cross_entropy is wrong as well.

Similar example in which the gradients differ as well.

```python
import torch
import torch.nn.functional as F
from liger_kernel.transformers.functional import liger_cross_entropy


def run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn):
    logits_p = F.softmax(logits_p, dim=-1)
    logits_p = logits_p.clone().detach().requires_grad_(True)
    p = F.softmax(logits_p, dim=-1)

    BT, V = p.shape
    q = torch.randint(0, V, (BT,), device="cuda", dtype=torch.long)

    cross_entropy_loss = cross_entropy_fn(p, q)
    cross_entropy_loss.backward()

    print(f"Cross Entropy Loss: {cross_entropy_loss.item()}")
    print(f"Input logits_p: {logits_p.sum()} (checksum)")
    print(f"Input logits_q: {logits_q.sum()} (checksum)")
    print(f"Gradients of logits_p (batch item 0): {logits_p.grad[0]}")


torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")


run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=F.cross_entropy)

print()
print("LIGER:")
run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=liger_cross_entropy)
```

**Comment by lancerts** [comment]
> @ByronHsu could you check this one quickly? If what I'm saying is true, cross_entropy, possibly fused_cross_entropy is wrong as well.
> 
> Similar example in which the gradients differ as well.
> 
> ```python
> import torch
> import torch.nn.functional as F
> from liger_kernel.transformers.functional import liger_cross_entropy
> 
> 
> def run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn):
>     logits_p = F.softmax(logits_p, dim=-1)
>     logits_p = logits_p.clone().detach().requires_grad_(True)
>     p = F.softmax(logits_p, dim=-1)
> 
>     BT, V = p.shape
>     q = torch.randint(0, V, (BT,), device="cuda", dtype=torch.long)
> 
>     cross_entropy_loss = cross_entropy_fn(p, q)
>     cross_entropy_loss.backward()
> 
>     print(f"Cross Entropy Loss: {cross_entropy_loss.item()}")
>     print(f"Input logits_p: {logits_p.sum()} (checksum)")
>     print(f"Input logits_q: {logits_q.sum()} (checksum)")
>     print(f"Gradients of logits_p (batch item 0): {logits_p.grad[0]}")
> 
> 
> torch.manual_seed(0)
> logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
> logits_q = torch.randn(8, 8, device="cuda")
> 
> 
> run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=F.cross_entropy)
> 
> print()
> print("LIGER:")
> run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=liger_cross_entropy)
> ```

```
import torch
import torch.nn.functional as F
from liger_kernel.transformers.functional import liger_cross_entropy


def run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn):
    logits_p = F.softmax(logits_p, dim=-1)
    logits_p = logits_p.clone().detach().requires_grad_(True)
    p = F.softmax(logits_p, dim=-1)

    BT, V = p.shape
    q = torch.randint(0, V, (BT,), device="cuda", dtype=torch.long)

    cross_entropy_loss = cross_entropy_fn(p, q)
    cross_entropy_loss.backward()

    print(f"Cross Entropy Loss: {cross_entropy_loss.item()}")
    print(f"Input logits_p: {logits_p.sum()} (checksum)")
    print(f"Input logits_q: {logits_q.sum()} (checksum)")
    print(f"Gradients of logits_p (batch item 0): {logits_p.grad[0]}")


torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")


run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=F.cross_entropy)

print()
print("LIGER:")

torch.manual_seed(0)
logits_p = torch.randn(8, 8, requires_grad=True, device="cuda")
logits_q = torch.randn(8, 8, device="cuda")

run_kl_div_experiment(logits_p, logits_q, cross_entropy_fn=liger_cross_entropy)
```
can you try this?

**Comment by S1ro1** [comment]
That one breaks for me as well, I think the same case, one softmax after detaching

**Comment by Tcc0403** [comment]
might be related to #254 ?

**Comment by Tcc0403** [comment]
I did a small test with a modified cross_entropy function that doesn't store gradients at input tensors, and it solved the problem.
So I guess some nn.function do store the ouput tensors for backporp, then those tensors are changed by liger_ce after, which caues an issue with gradient computation.

**Comment by Tcc0403** [comment]
[In-place correctness checks](https://github.com/linkedin/Liger-Kernel/pull/262#issuecomment-2374924404)
I think doing inplace operations in triton kernel makes pytorch lost track of it.

**Review by lancerts — APPROVED** [review]
